xof.md

```diff
@@ -0,0 +1,206 @@
+- finalize may be done as part of the first squeeze operation
+- squeeze can be called multiple times.
+
+OpenSSL XOF Requirements
+------------------------
+
+The current OpenSSL implementation of XOF only supports a single call to squeeze.
+The assumption exists in both the high level call to EVP_DigestFinalXOF() as well
+as in the lower level SHA3_squeeze() operation (Of which there is a generic c version,
+as well as assembler code for different platforms).
+
+A decision has to be made as to whether a new API is required, as well as considering
+how the change may affect existing applications.
+The changes introduced should have a minimal affect on other related functions that
+share the same code (e.g SHAKE and SHA3 share functionality).
+Older providers that have not been updated to support this change should produce an
+error if a newer core is used that supports multiple squeeze operations.
+
+API Discussion of Squeeze
+-------------------------
+
+### Squeeze
+
+Currently EVP_DigestFinalXOF() uses a flag to check that it is only invoked once.
+It returns an error if called more than once. When initially written it also did a 
+reset, but that code was removed as it was deemed to be incorrect.
+
+If we remove the flag check, then the core code will potentially call low level squeeze code
+in a older provider that does not handle returning correct data for multiple calls.
+To counter this the provider needs a mechanism to indicate that multiple calls are allowed.
+This could just be a new gettable flag (having a separate provider function should not be necessary).
+
+#### Proposal 1
+
+Change EVP_DigestFinalXOF(ctx, out, outlen) to handle multiple calls.
+Possibly have EVP_DigestSqueeze() just as an alias method?
+Changing the code at this level should be a simple matter of removing the flag check.
+
+##### Pros
+
+  - New API is not required
+
+##### Cons
+
+  - Final seems like a strange name to call multiple times.
+
+#### Proposal 2
+
+Keep EVP_DigestFinalXOF() as a one shot function and create a new API to handle the
+multi squeeze case e.g.
+```
+EVP_DigestSqueeze(ctx, out, outlen).
+```
+
+##### Pros
+
+  - Seems like a better name.
+
+##### Cons
+
+  - Adds an extra API.
+  - The interaction between the 2 API's needs to be clearly documented.
+     - A call to EVP_DigestSqueeze() after EVP_DigestFinalXOF() would fail.
+     - A call to EVP_DigestFinalXOF() after the EVP_DigestSqueeze() would fail? Is this confusing.
+
+API Discussion of other XOF API'S
+---------------------------------
+
+### Init
+
+The digest can be initialized as normal using:
+```
+md = EVP_MD_fetch(libctx, "SHAKE256", propq);
+ctx = EVP_MD_CTX_new();
+EVP_DigestInit_ex2(ctx, md, NULL);
+```
+
+### Absorb
+
+Absorb can be done by multiple calls to:
+```
+EVP_DigestUpdate(ctx, in, inlen);
+```
+
+#### Proposal:
+
+Do we want to have an Alias function?
+```
+EVP_DigestAbsorb(ctx, in, inlen);
```

Comment:
I'd say no. It doesn't seem to add anything.

Reply:
Maybe not - I used it for testing..